gluetool/result.py

```diff
@@ -0,0 +1,183 @@
+# Type annotations
+# pylint: disable=unused-import,wrong-import-order
+from typing import cast, Any, Generic, Optional, TypeVar, Union  # noqa
+
+
+#
+# Result and its handling.
+#
+#
+# A ``Result`` can be either ``Ok(value)`` - valid result, contains a meaningful
+# value - or ``Error(error)`` which represents an error, carrying error's description.
+#
+
+# T represents the type of valid value...
+T = TypeVar("T")
+# ... and E represents type of the error description.
+E = TypeVar("E")
+
+
+class Result(Generic[T, E]):
+    """
+    A simple `Result` type inspired by Rust.
+
+    A ``Result`` can be either ``Ok(value)`` - valid result, contains a meaningful
+    value - or ``Error(error)`` which represents an error, carrying error's description.
+
+    :param bool _is_ok: ``True`` when the ``_value`` is OK-ish.
+    :param _value: the value carried by the result.
+    :param bool _force: guards against accidental direct use.
+    """
+
+    def __init__(self, _is_ok, _value, _force=False):
+        # type: (bool, Union[T, E], bool) -> None
+        """
+        .. warning::
+
+           Do not instantiate ``Result`` instrances directly, **always** use either
+           :py:func:`Ok` or :py:func:`Error` functions. Othwerise, type guarantees
+           cannot be given.
+        """
+
+        if not _force:
+            raise RuntimeError('Do not instantiate Result objects directly.')
+
+        self._is_ok = _is_ok
+        self._value = _value
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        # pylint: disable=protected-access
+
+        return (self.__class__ == other.__class__ and
+                self.is_ok == cast(Result, other).is_ok and
+                self._value == other._value)
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+
+        return not self == other
+
+    def __hash__(self):
+        # type: () -> int
+
+        return hash((self.is_ok, self._value))
+
+    def __repr__(self):
+        # type: () -> str
+
+        if self.is_ok:
+            return 'Ok({})'.format(repr(self._value))
+
+        return 'Err({})'.format(repr(self._value))
```

Comment:
I would use ```Error``` here, not Err, to be consistent

Reply:
Fixed.